src/mlpack/methods/reinforcement_learning/replay/hindsight_replay.hpp

```diff
@@ -0,0 +1,423 @@
+   * experience replay in replay buffer
+   * @param nSteps Number of steps to look in the future.
+   * @param dimension The dimension of an encoded state.
+   * @param strategy goal selection startegy for HER
+   */
+  HindsightReplay(const size_t batchSize,
+                    const size_t capacity,
+                    const size_t herRatio = 4,
+                    EnvironmentType environment = EnvironmentType(),
+                    goalStrategy strategy = goalStrategy::FUTURE,
+                    const size_t nSteps = 1,
+                    const size_t dimension = StateType::dimension) :
+      batchSize(batchSize),
+      capacity((herRatio+1)*capacity),
+      position(0),
+      full(false),
+      herRatio(herRatio),
+      goalSelectionStrategy(strategy),
+      nSteps(nSteps),
+      states(dimension, (herRatio+1)*capacity),
+      actions((herRatio+1)*capacity),
+      rewards((herRatio+1)*capacity),
+      nextStates(dimension, (herRatio+1)*capacity),
+      goals(dimension, (herRatio+1)*capacity),
+      isTerminal((herRatio+1)*capacity),
+      environment(environment)
+  { /* Nothing to do here. */ }
+
+  /**
+   * Sample goal according to goal strategy
+   * 
+   * @param transitionIndex index of current transition
+   */
+  void SampleGoal(StateType& desiredGoal, int transitionIndex)
+  {
+    if (goalSelectionStrategy == goalStrategy::FINAL)
+    {
+      // Sample goal as the final state of the epsiode
+      desiredGoal = episodeTransitions[episodeTransitions.size() -1].nextState;
+    }
+    else if (goalSelectionStrategy == goalStrategy::FUTURE)
+    {
+      // Sample goal as random state from that transition to end of epsiode
+      size_t index = arma::randi(
+        arma::distr_param(transitionIndex+1, episodeTransitions.size() -1));
+      desiredGoal = episodeTransitions[index].nextState;
+    }
+    else if (goalSelectionStrategy == goalStrategy::EPISODE)
+    {
+      // Sample goal as random state from the whole epsiode
+      desiredGoal = episodeTransitions[arma::randi(
+        arma::distr_param(0, episodeTransitions.size() -1))].nextState;
+    }
+  }
+
+  /**
+   * Sample goals according to goal strategy and HER ratio
+   * 
+   * @param desiredGoals desired goals as per goal strategy
+   * @param  transitionIndex index of current transition
+   */
+  void SampleGoals(std::vector<StateType>& desiredGoals, int transitionIndex)
+  { 
+    StateType desiredGoal;
+
+    // Store HER transitions based on strategy for each transition in epsiode
+    for (size_t goalIndex = 0; goalIndex < herRatio; ++goalIndex)
+    {
+      SampleGoal(desiredGoal, transitionIndex);
+
+      desiredGoals.push_back(desiredGoal);
+    }
+  }
+
+  /**
+   * Sample goal according to goal strategy
+   * @param discount The discount parameter.
+   */
+  void StoreHERTransitions(const double& discount)
+  { 
+    std::vector<Transition> baseTransitions = episodeTransitions;
```

Comment:
Do we need `episodeTransitions` after this? If not, then ig it's fine to corrupt its value here.

Reply:
Why don't we use `episodeTransitions` directly?

Reply:
I think we can now ! Previously i had done this to accomodate something else but looks like i have changed that. Since i am storing HER Transitions only just before termination of episode.

Reply:
> Why don't we use `episodeTransitions` directly?

we actually need it ! when we future strategy we want to sample goal from transitions which happened after the required transition. So if pop_back in `epsiodeTransitions` when i sample a index using randi i get index out of bounds issue.